Rubberduck.Core/UnitTesting/TestEngine.cs

```diff
@@ -21,17 +22,43 @@ public class TestEngine : ITestEngine
         private readonly RubberduckParserState _state;
         private readonly IFakesFactory _fakesFactory;
         private readonly IVBETypeLibsAPI _typeLibApi;
+        private readonly IUiDispatcher _uiDispatcher;
 
         private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
 
-        public TestEngine(TestExplorerModel model, IVBE vbe, RubberduckParserState state, IFakesFactory fakesFactory, IVBETypeLibsAPI typeLibApi)
+        private bool _testRequested;
+        private IEnumerable<TestMethod> _tests;
+
+        public TestEngine(TestExplorerModel model, IVBE vbe, RubberduckParserState state, IFakesFactory fakesFactory, IVBETypeLibsAPI typeLibApi, IUiDispatcher uiDispatcher)
         {
             Debug.WriteLine("TestEngine created.");
             Model = model;
             _vbe = vbe;
             _state = state;
             _fakesFactory = fakesFactory;
             _typeLibApi = typeLibApi;
+            _uiDispatcher = uiDispatcher;
+
+            _state.StateChanged += StateChangedHandler;
+        }
+
+        private void StateChangedHandler(object sender, ParserStateEventArgs e)
+        {
+            if (_testRequested && (e.State == ParserState.Ready))
+            {
+                _testRequested = false;
+                _uiDispatcher.InvokeAsync(() =>
+                {
+                    RunInternal(_tests);
+                });
+            }
+
+            if (_testRequested && (e.State == ParserState.ResolverError || 
+                                   e.State == ParserState.Error ||
+                                   e.State == ParserState.UnexpectedError))
```

Comment:
we should be able to unify that check into a single `e.State.IsError()`. That can be used across the codebase for consistent handling of all error states, even those we introduce in the future...

Reply:
Good idea. I made it `e.IsError`, though, since `e.State` is an enum.